RFC 0010 - Expressions as Constraints (#280)
jvalue/jayvee: rfc/0010-constraint-expressions/README.md

## Motivation

Defining constraints is currently very bloated on the syntax level.

Example:
```
constraint HundredScale oftype RangeConstraint {
  lowerBound: 1;
  lowerBoundInclusive: false;
  upperBound: 100;      
}
```

## Explanation

Instead of using this block-like syntax, we can use expressions to define constraints.

The general syntax looks like this: 
```
constraint <name> on <primitive value type> = <expression with parameter value, evaluating to boolean>
```

### Examples

The following subsections will go over the existing constraints and showcase them in an example.

#### AllowlistConstraint

```
constraint TimeUnitString on text = 
  value =="ms" or 
  value == "s" or 
  value == "min" or 
  value == "h" or 
  value == "d" or 
  value == "m" or 
  value == "y";
```

#### DenylistConstraint
```
constraint NoPrimaryColors on text = 
  not (value == "red" or value == "blue" or value == "yellow");

// alternative
constraint NoPrimaryColors on text = 
  value != "red" and 
  value != "blue" and 
  value != "yellow";
```

#### LengthConstraint
```
constraint JavaStringLength on text = value.length  >= 0 and value.length  <= maxLength;
```

#### RangeConstraint
```

> constraint HundredScale on decimal = value > 1 and value <= 100; 

Reviewer: I agree on not doing it here but I want to mention that I consider that syntax a big plus for intuitive use. A value between two limits comes up surprisingly often and is weird to express in classic boolean checks.

> constraint HundredScale on decimal = value > 1 and value <= 100; 

Author: => https://github.com/jvalue/jayvee/issues/288

```

#### RegexConstraint
```
constraint IFOPT_Format on text = value matches /[a-z]{2}:\d+:\d+(:\d+)?(:\d+)?/;
```

### Hidden Enhancements
* Expressions allow parameters (keyword `value`).
* The length of string parameters is accessible by `value.length`.

> * The assignment operator `=` instead of `:` because the latter looks off in that context.

Author: Yes, probably not a change to do in disguise in this PR.

> * The assignment operator `=` instead of `:` because the latter looks off in that context.

Reviewer: I dislike using `=` here, I'd go with `:` in the context of this RFC. If we want to switch to `=` we should do so in a separate decision I think.

> * The binary string operator `matches` is introduced.

Reviewer 2: ```suggestion
* The binary operator `matches` is introduced which evaluates to a boolean value. The usage is `<text> matches <regex>`.
```

### Usage in ValueTypes

> In ValueTypes, these constraints can be added like before by name. Combining constraints on non-matching value types leads to an error at runtime in the IDE hints.

Reviewer 2: Reporting such errors at runtime is unnecessarily late, it can be done statically by the language server:
```suggestion
In ValueTypes, these constraints can be added like before by name. Combining constraints on non-matching value types leads to a diagnostic error at compile-time (by the language server).
```

```

> valuetype GasFillLevel oftype number {

Reviewer 2: ```suggestion
valuetype GasFillLevel oftype decimal {
```

    constraints: [ GasFillLevelRange ];
}
```

> Alternatively, constraints can be defined inline with the caveat that they are not reusable elsewhere. Value then adheres to the type of the ValueType.

Reviewer 2: ```suggestion
Alternatively, constraints can be defined inline with the caveat that they are not reusable elsewhere. Then the value, that the `value` keyword refers to, adheres to the type of the enclosing ValueType.
```

```

> valuetype GasFillLevel oftype number {

Reviewer 2: ```suggestion
valuetype GasFillLevel oftype decimal {
```

    constraints: [ value >= 0 and value <= 100 ];
}
```

## Drawbacks

- Inconsistency of constraint definition via `=` and property initialization via `:`.

> - Future constraints have to be expressible by a boolean expression.

Reviewer 2: Not necessarily if we also keep the current concept for configurable constraints. It would be well suited for more complex / builtin constraints (as elaborated above).

> - Users now have to be able to write expressions instead of just supplying property values.

Reviewer 2: Will be solved via #273 and is not a drawback in my opinion.

## Alternatives

- Syntax to define parameter in constraint could be different, e.g., `constraint MyConstraint = (number) => <expression>`.

> - Allow if/else statements in a more imperative way rather than packing everything into one expression.

Reviewer 2: I think this could indeed be beneficial for cases where the expression becomes large and confusing. An alternative would be to nest constraints, so the overall complexity could be broken up into smaller sub-constraints.

As this is not an exclusive alternative, I'd shift this point to the `Possible Future Changes/Enhancements` section.

For now, I'd just go with the boolean expressions suggested in this RFC and observe whether the need for splitting up complex expressions arises at some point.

- Put the expression into a `{}`-wrapped scope for consistency (probably together with if/else statements, looks more like a method).

## Possible Future Changes/Enhancements

> - Allow renaming of `value` to a custom parameter name, e.g., `for text as x`.

Reviewer 2: Do you have a use case in mind where this would be beneficial?

Author: Sure. I think it can be beneficial for the code's readability as custom parameter names are more expressive while  `value` stay generic. Second use-case related to composite value types where - depending on how we will implement that - having multiple named parts could be referable in constraints. Not problematic if we require to prepend the `value.` for dereferencing, though.

Reviewer: I'd go with value and dereferencing for now but I can follow your reasoning.

> - add `in` operator for collections for syntactic sugar.

Reviewer 2: I think this is generally a good idea and it should be incorporated in this RFC. An `in` operator would make it easier to define constraints similar to the current `AllowlistConstraint` and `DenylistConstraint`:
```jayvee
constraint NoPrimaryColors on text =
  not (value in ["red", "blue", "yellow"]);
```

Author: => https://github.com/jvalue/jayvee/issues/289

> - Allow users to define own `ConstraintTypes` that allow the current syntax again.

Reviewer 2: In case we decide to also keep our current concept, this would be a nice way to incorporate the two approaches (boolean expressions on the one hand and configurable constraint types on the other).
